editor/state/save-state.js

```diff
@@ -0,0 +1,72 @@
+/**
+ * External dependencies
+ */
+import { combineReducers } from 'redux';
+
+/**
+ * Internal dependencies
+ */
+import { getCurrentPost, hasEditorUndo } from '../selectors';
+
+/**
+ * Action creator returning an action object used in signalling that the post's
+ * dirty saved state should be toggled.
+ *
+ * @param  {Boolean} isDirty Whether post is dirty
+ * @return {Object}          Action object
+ */
+export function toggleDirty( isDirty ) {
+	return {
+		type: 'TOGGLE_DIRTY',
+		isDirty,
+	};
+}
+
+/**
+ * Redux middleware for save state dirtying tracking.
+ *
+ * @param  {Function} store Redux store object
+ * @return {Function}       Middleware function
+ */
+export const middleware = ( store ) => ( next ) => ( action ) => {
+	const state = store.getState();
+	const result = next( action );
+	const nextState = store.getState();
+
+	let isDirty;
+	if ( getCurrentPost( nextState ) !== getCurrentPost( state ) ) {
+		// Current post reflects last known post save, so if these references
+		// differ, we can assume that the post has been saved
+		isDirty = false;
+	} else if ( state.editor !== nextState.editor && hasEditorUndo( nextState ) ) {
+		// Editor state tracks post attribute edits, block order, block
+		// attributes. If any of these change, assume the post is dirtied.
+		// Without undo indicates an editor history reset (initial load).
+		isDirty = true;
+	}
+
+	// Avoid dispatch except when we know that the state is changing
+	if ( undefined !== isDirty && isDirty !== nextState.saveState.isDirty ) {
+		store.dispatch( toggleDirty( isDirty ) );
+	}
+
+	return result;
+};
```

Comment:
I'm wondering why this needs to be a middleware. Can't we do this simply with a "reducer enhancer":

```js
enhancedEditorReducer( state, action ) {
	const newState = editorReducer( state, action );
	const isDirty = some( newState, ( value, key ) => key !== 'dirty' && value !== state[ key ] );
	if ( isDirty !== newState.dirty ) {
		newState.dirty = isDirty;
	}

	return newState;
}
```

Reply:
I like this idea, and it sounds like a nice simplification. It would mean that the dirty flag would become part of editor state subtree, but I think that's reasonable.

Reply:
Actually, I think I missed something here, When do we reset the isDirty to false. I think this enhancer needs to be in the higher level to take into account resetting the `dirty` value when the currentPost changes

Reply:
Right, unless we could rely on specific action types as signals for reset (`RESET_POST`, `UPDATE_POST`), then the reducer enhancer for editor could simply listen for those.

Also in your original example, ideally we don't need the `some` and could do a strict equality check `newState !== state`.